### PR TITLE
Add more type checks for service call arguments

### LINKED ISSUE #336: Service need to handle object types
https://github.com/storyscript/platform-engine/blob/ef77bd81cdc7dbaad9e45c58392924d739de29d0/asyncy/processing/Services.py#L298-L303

OMG supports complex `object` types.

see also: https://github.com/storyscript/storyscript/issues/940

### LINKED ISSUE #570: Fix smart_insert, raise_for_type_mismatch
Both use `command_conf.get('type')` but `type` will be found in `command_conf[argument_name]`.

See [here](https://github.com/storyscript/runtime/blob/2ebffa90a0764ff2760bbf22c72fd900476b1c0c/storyruntime/processing/Services.py#L423) and [here](https://github.com/storyscript/runtime/blob/2ebffa90a0764ff2760bbf22c72fd900476b1c0c/storyruntime/processing/Services.py#L383).

### LINKED ISSUE #571: Add number range and string pattern checks for service arguments
https://github.com/storyscript/runtime/blob/ef77bd81cdc7dbaad9e45c58392924d739de29d0/asyncy/processing/Services.py#L298-L326

OMS supports `range` specification for `int`, `float` type arguments and `pattern` specification for `string` type arguments. See https://microservice.guide/schema/actions/#arguments

### PR DESCRIPTION
**Pull request by anukul**
Closes #336 
Closes #570 
Closes #571

### CODE REVIEW

**Comment by anukul** [comment]
Do we want to support range check on floats?
I think the current implementation (`<=`) might break for appropriate values.

**Comment by judepereira** [comment]
Merge away.

File: tests/unit/processing/Services.py [hunk 1]
```diff
@@ -201,34 +201,348 @@ def test_smart_insert(patch, story, value):
...
+}, {
+    # int, max, invalid
+    'arg_conf': {
+        'type': 'int',
+        'range': {
+            'max': 20,
+        }
+    },
+    'value': 25,
+    'valid': False
```
[the remark this answers is not in the document]
Comment (wilzbach) [reply tests/unit/processing/Services.py]: Other interesting cases are where range isn't set or an empty dict.

File: tests/unit/processing/Services.py [hunk 2]
```diff
@@ -201,34 +201,348 @@ def test_smart_insert(patch, story, value):
...
+                                             case['value'], case['arg_conf'])
+
+
+@mark.parametrize('case', [{
+    'pattern': '[a-z]+',
+    'value': 'abc',
+    'valid': True
+}, {
+    'pattern': '[a-z]+',
+    'value': 'abc123',
```
[the remark this answers is not in the document]
Comment (anukul) [reply tests/unit/processing/Services.py]: I'm assuming empty pattern should only allow empty strings to pass.
For invalid patterns, what should the unit test assert?
In practice, the exception will be caught on its way up.

File: tests/unit/processing/Services.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (anukul) [reply tests/unit/processing/Services.py]: Added tests for `value: NaN`, `range: {}`.
Range unset is already covered, and `range: NaN` behavior is currently similar to invalid pattern. (exception thrown, caught elsewhere)

[the remark this answers is not in the document]
Comment (wilzbach) [reply tests/unit/processing/Services.py]: > Range unset is already covered, and range: NaN behavior is currently similar to invalid pattern. (exception thrown, caught elsewhere)

Cool! Do we have tests for this? Is this done on startup?

File: tests/unit/processing/Services.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (wilzbach) [reply tests/unit/processing/Services.py]: > I'm assuming empty pattern should only allow empty strings to pass.

No, it should match everything. See e.g. https://2ality.com/2012/09/empty-regexp.html
The same should apply to `pattern: None` (there's even no need to run the regex).

> For invalid patterns, what should the unit test assert?
> In practice, the exception will be caught on its way up.

Well, I would love to see that. An ideal _integration_ test would be to throw a dummy service with such a JSON specification against the runtime and assert that this exception does get thrown.
Anyhow, as OMS hopefully disallows this, it's not very important.

[the remark this answers is not in the document]
Comment (anukul) [reply tests/unit/processing/Services.py]: Added empty pattern

File: tests/unit/processing/Services.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (anukul) [reply tests/unit/processing/Services.py]: No integration tests for services yet, this is caught in `Lexicon.execute_line`

[the remark this answers is not in the document]
Comment (wilzbach) [reply tests/unit/processing/Services.py]: -> https://github.com/storyscript/runtime/issues/579

**Pull request merged**